### DISCUSSION TITLE
Help! Format on Save in HTML - Closing Tags

### DISCUSSION THREAD

**Discussion by mnmalist** [zed-industries/zed#18225] [Feature Requests]
In HTML, when Format on Save is enabled, HTML tags that are closing with `>` are being reformatted to XHTML style `/>`.

For example, if the code is written like this:

```
<link rel="stylesheet" href="style.css">
```

it will always reformat to be closed XHTML style ('/>'), like this:

```
<link rel="stylesheet" href="style.css" />
```

How can I fix this, is there an option to disable adding `/` when closing tags?

I still want to use format, it's just this one issue I have a problem with.
Thanks!

**Comment by SomeoneToIgnore** [comment]
Assuming that the language in question is HTML (? or something else frontend-related), most probably you're using `prettier` if you have not disabled it explicitly.

Maybe https://prettier.io/docs/en/options.html or some GitHub issues contain the related setting?

**Reply by mnmalist** [reply]
Yes, the language is HTML, and the default formatter in Zed is being used (not sure which one it is). On the link that you provided, all of the examples state that valid HTML5 is in use. Maybe it's something related to HTML extension? I really don't know, but it's forcing XHTML style for closing tags.

**Reply by SomeoneToIgnore** [reply]
> not sure which one it is

Let's assume it's prettier.
Take that file away, do something like `mkdir test && cd test && npm i --include=dev --save prettier` and `node ./node_modules/prettier/bin/prettier.cjs test.html` to see how it formats things.

Then you can set `"prettier": {}` in the newly created `package.json` and try to find a configuration that does something else (if it even exists).

> all of the examples state that valid HTML5 is in use

Well, isn't the most valid HTML5 is a `<link ..  />` one? 
Prettier is pretty opinionated on how things are formatted.

-------------

Alternatively, forbid prettier and use LSP formatting as `"formatter": "language_server"` in the corresponding language section in the config, maybe it will do something differently.

**Reply by mnmalist** [reply]
In `zed/extensions/html/languages/html/indents.scm` I found the code:
```
(self_closing_tag "/>" @end) @indent
```

Maybe if it would be possible to override this in settings somehow, and just delete the `/` character? 

Any ideas?

**Reply by sylbru** [reply]
> Well, isn't the most valid HTML5 is a <link ..  /> one?

Just answering on this: explicitly self-closing tags are a XHTML thing, and virtually nobody is or has been really using it (as much as I’d have loved to!), so I’d argue it’s more "correct" without the slash (but I think both are valid). A very good read about this issue: https://jakearchibald.com/2023/against-self-closing-tags-in-html/

**Comment by notpeter** [comment]
Zed ships with two formatters for html out of the box: prettier (default) and `html-language-server`.  Neither of which support the custom formatting you desire.  Zed does not make an html formatter, but if you hate `/` enough to not use prettier you are free to disable `format_on_save`:  

You can disable format on save:
```
  "languages": {
    "HTML": {
      "format_on_save": "off"
    },
  }
```

Zed also supports [Custom Formatters](https://zed.dev/docs/configuring-zed#formatter) if you have a program that applies the formatting style you're looking for.  Otherwise this is strictly an upstream issue.

**Reply by mnmalist** [reply]
@notpeter  That's exactly what I did! 

I tired to use `html-language-server` before disabling, not sure if I got it right:
```
"languages": {
    "HTML": {
      "formatter": "html-language-server"
    },
  }
  ```
  
  But it made no difference.
  
  Thanks for clarification!

**Reply by notpeter** [reply]
These are the settings you need for that:
```
  "languages": {
    "HTML": {
      "prettier": {
        "allowed": false
      },
      "formatter": "language_server"
    }
  }
```

But it is not currently working. I looked in the LSP logs and we're not sending the `textDocument/formatting` event to the language server so something is off.  Unrelatedly I've been doing some `html-language-server` adjacent thing so I'll see if I can figure out what's going on.

P.S. Apologies for being a bit smug above.
